### PR TITLE
0.10bounds

### PR DESCRIPTION
**Pull request by MrChico**
Adds bounds checking to some key calculations in `get_attestation_deltas`, in the spirit of @daejunpark s [bounds-check branch](https://github.com/runtimeverification/beacon-chain-spec/tree/bounds-check). This also adds the bounds checking as a target for ci so that it is run on every PR

### CODE REVIEW

File: beacon-chain.k [hunk 1]
```diff
@@ -1398,12 +1406,16 @@ def get_attestation_deltas(state: BeaconState) -> Tuple[Sequence[Gwei], Sequence
                        UnslashedIndices,
                        AttestingBalance,
                        #mapMapPair(Rewards => #if contains(I, UnslashedIndices)
-                                              #then Rewards[I <- ({Rewards[I]}:>Int +Int get_base_reward(I)
-                                                                  *Int AttestingBalance /Int TotalBalance)]
+                                              #then Rewards[I <- ei(0, 2 ^Int 64 -Int 1, ci({Rewards[I]}:>Int) +UInt64 get_base_reward(I)
```
Comment (daejunpark) [inline beacon-chain.k]: I think `ci({Rewards[I]}:>Int)` should be changed to `ui(MIN, MAX, {Rewards[I]}:>Int)`, with proper MIN/MAX bounds.  The thing is that `ci(_)` is meant to be used only for constants, but `Rewards[I]` is not a constant.  I guess the bounds can be given the same bounds of `get_base_reward(I)`?

File: beacon-chain.k [hunk 2]
```diff
@@ -1398,12 +1406,16 @@ def get_attestation_deltas(state: BeaconState) -> Tuple[Sequence[Gwei], Sequence
...
+                                                                  *UInt64 ( ui(0, TOTAL_ETHER_SUPPLY, AttestingBalance)
+                                                                  /UInt64 ci(EFFECTIVE_BALANCE_INCREMENT))
+                                                                  /UInt64 ( ui(EFFECTIVE_BALANCE_INCREMENT, TOTAL_ETHER_SUPPLY, TotalBalance)
+                                                                  /UInt64 ci(EFFECTIVE_BALANCE_INCREMENT)))
+                                                                  ]
                                               #else Rewards
                                               #fi,
                                   Penalties => #if notBool contains(I, UnslashedIndices)
-                                               #then Penalties[I <- ({Penalties[I]}:>Int +Int get_base_reward(I))]
+                                               #then Penalties[I <- ei(0, 2 ^Int 64 -Int 1, (ci({Penalties[I]}:>Int) +UInt64 get_base_reward(I)))]
```
Comment (daejunpark) [inline beacon-chain.k]: Same here: `ci({Penalties[I]}:>Int)` should be changed to `ui(MIN, MAX, {Penalties[I]}:>Int)`?

File: beacon-chain.k [hunk 3]
```diff
@@ -1511,7 +1523,7 @@ def get_attestation_deltas(state: BeaconState) -> Tuple[Sequence[Gwei], Sequence
   // penalties[index] += Gwei(BASE_REWARDS_PER_EPOCH * get_base_reward(state, index))
   syntax Map ::= updateBasePenalties(Map, Int) [function]
   rule updateBasePenalties(Penalties, INDEX)
-       => Penalties[INDEX <- {Penalties[INDEX]}:>Int +Int (BASE_REWARDS_PER_EPOCH *Int get_base_reward(INDEX))]
+       => Penalties[INDEX <- ei(0, 2 ^Int 64 -Int 1, ci({Penalties[INDEX]}:>Int) +UInt64 (ci(BASE_REWARDS_PER_EPOCH) *UInt64 get_base_reward(INDEX)))]
```
Comment (daejunpark) [inline beacon-chain.k]: Same here for `ci({Penalties[INDEX]}:>Int)`.

File: beacon-chain.k (hunk 1 above)
Comment (daejunpark) [reply beacon-chain.k]: One way to give the same bounds of `get_base_reward(...)` to `Rewards[I]` is that you come up with a reasonable MIN/MAX bounds, and ensure that `get_base_reward(...)` indeed falls within the bounds, like:
```
rule get_base_reward(...) => ensureBounds(MIN, MAX, ...)
```
where `ensureBounds` is similar to `ei` but preserves the UInt64 value.
```
syntax UInt64 ::= ensureBounds(Int, Int, UInt64) [function]
rule ensureBounds(L, U, #uint64(L0, U0, V)) => #uint64(L0, U0, V) requires L <=Int L0 andBool U0 <=Int U
```

Comment (MrChico) [reply beacon-chain.k]: I agree that `Rewards[I]` should be given bounds. But I don't think the appropriate bound would be the same as get_base_reward.
There seems to be an error in the k specs here, which was partly present in the [python specs](https://github.com/ethereum/eth2.0-specs/pull/1682/files) as well, in that we should be adding to the rewards for every attestation (right now we are resetting it).

With the correct implementation, it seems that the appropriate max bound for Rewards would need to be:
```
get_base_reward(state, index) * (attesting_balance // total_balance) * (ATTESTATIONS - 1)
```

where ATTESTATIONS would be the upper bound of `length(matching_source_attestations) + length(matching_target_attestations) + length(matching_head_attestations)` which is bounded to `3 * MAX_ATTESTATIONS * SLOTS_PER_EPOCH`

**Pull request merged**